### ISSUE TITLE
encode code-model and stack-alignment in IR

### ISSUE THREAD

**Issue by nickdesaulniers** [ClangBuiltLinux/linux#1377]
follow up to #1374 .

In https://lore.kernel.org/lkml/20210518190106.60935-1-nathan@kernel.org/ @nathanchance sent a patch for `-plugin-opt=-code-model=kernel`, and `-plugin-opt=-stack-alignment=`. We should encode these in LLVM's IR, and drop them from Kbuild.

**Comment by nickdesaulniers** [comment]
stack alignment: https://reviews.llvm.org/D103048

**Comment by nickdesaulniers** [comment]
`-mcmodel=kernel` corresponds to `-plugin-opt=-code-model=kernel`, IIUC, though setting `-mcmodel=kernel` should result in the following module level metadata:

```llvm
!17 = !{i32 1, !"Code Model", i32 2}
```
(verified back to llvm-9)

So I don't think it's necessary to re-specify `-plugin-opt=-code-model=kernel` for LTO builds of the kernel. 0024430e920f shows that it wasn't being used for a while anyways.  @samitolvanen , do you recall a specific case where it was needed?  I suspect we can drop its use outright from the kernel.  If so, when I send a kernel patch removing `-plugin-opt=-stack-alignment=` (once https://reviews.llvm.org/D103048 lands), then I can remove its use in the same patch.

---

Also, https://git.kernel.org/pub/scm/linux/kernel/git/kees/linux.git/commit/?h=for-linus/clang/features demonstrates that `-warn-stack-size=` demonstrates a problem similar to stack alignment.

**Comment by nickdesaulniers** [comment]
`-warn-stack-size=` will be tough to support since it's consumed by PrologEpilogInserter which consumes MIR. Maybe this would need to be represented in LLVM IR, then in MachineModuleInfo.  Tests to ensure pass with such a change:
```
Failed Tests (5):
  Clang :: Frontend/backend-diagnostic.c
  Clang :: Misc/backend-stack-frame-diagnostics-fallback.cpp
  Clang :: Misc/backend-stack-frame-diagnostics.cpp
  LLVM :: CodeGen/ARM/warn-stack.ll
  LLVM :: CodeGen/X86/warn-stack.ll
```

**Comment by nickdesaulniers** [comment]
> then in MachineModuleInfo

ah, even easier; MachineFunction's retain a reference to their LLVM IR `Function` (accessible via `MachineFunction::getFunction()`), so we can just get the IR metadata from that easily.

**Comment by samitolvanen** [comment]
> @samitolvanen , do you recall a specific case where it was needed? I suspect we can drop its use outright from the kernel. 

I don't remember, I think it came from @gwelymernans.

**Comment by bwendling** [comment]
That option might be a hold over from the gold plugin days. It was needed for that I think.

**Comment by nickdesaulniers** [comment]
warn frame size: https://reviews.llvm.org/D103928

**Comment by torvic9** [comment]
Hi,
as a noob, may I ask whether I could work around this with the following diff? 
Or is it too "strict" and we still need -code-model?
Thanks!

```diff
diff --git a/arch/x86/Makefile b/arch/x86/Makefile
index 307529417021..c0b863d2ab4a 100644
--- a/arch/x86/Makefile
+++ b/arch/x86/Makefile
@@ -200,9 +200,11 @@ endif
 KBUILD_LDFLAGS += -m elf_$(UTS_MACHINE)
 
 ifdef CONFIG_LTO_CLANG
+ifeq ($(shell test $(CONFIG_LLD_VERSION) -lt 130000; echo $$?),0)
 KBUILD_LDFLAGS	+= -plugin-opt=-code-model=kernel \
 		   -plugin-opt=-stack-alignment=$(if $(CONFIG_X86_32),4,8)
 endif
+endif
 
 ifdef CONFIG_X86_NEED_RELOCS
 LDFLAGS_vmlinux := --emit-relocs --discard-none
-- 
2.32.0
```

**Comment by nathanchance** [comment]
We can probably just drop `-plugin-opt=-code-model=kernel` then make `-plugin-opt=-stack-alignment=$(if $(CONFIG_X86_32),4,8)` conditional on `$(CONFIG_LLD_VERSION) -lt 130000` like you did.

**Comment by nickdesaulniers** [comment]
> That option might be a hold over from the gold plugin days. It was needed for that I think.

Huh, we didn't use `-plugin-opt=-code-model=kernel` in pixel 3 kernel sources, (though I do see `-plugin-opt=-code-model=large` in arch/arm64/Makefile added to `KBUILD_LDFLAGS_MODULE`; that has the comment `# Code model is not stored in LLVM IR, so we need to pass it also to LLVMgold`).  Interesting. Ah, [llvm commit 3dea3f9e0a7e2c72ff7e109d1581b62365fc7ae8 ("Pass code-model through Module IR to LTO which will use it.")](https://github.com/llvm/llvm-project/commit/3dea3f9e0a7e2c72ff7e109d1581b62365fc7ae8) by @cmtice fixed `-plugin-opt=-code-model=kernel` in clang-9: https://reviews.llvm.org/D52322.  Since we only support clang-10 and newer, we can indeed drop `-plugin-opt=-code-model=kernel` outright.

---
@torvic9 this would be a great kernel patch for you to send!

You might want to wait until https://reviews.llvm.org/D103928 is actually merged, but you can prep a patch here.  The same patch should additionally modify the use of `-plugin-opt=-warn-stack-size=` in the top level Makefile conditional on lld-13 as well.

You can add link tags for:
```
Link: https://github.com/ClangBuiltLinux/linux/issues/1377
Link: https://reviews.llvm.org/D52322
Link: https://reviews.llvm.org/D103048
Link: https://reviews.llvm.org/D103928
```
in the commit message, I would appreciate that.

I'm hoping to land https://reviews.llvm.org/D103928 today, so it would be good to have a kernel patch ready to go soon, since our CI is currently red without such kernel patch. (or it should be, let me find out what's going on, we should have had something somewhere fail).

**Comment by nathanchance** [comment]
Can confirm there is an error with ToT LLVM now. Sending that patch with `Cc: stable@vger.kernel.org` will be needed so 5.12 can be built with LLVM 13 too.

```
ld.lld: error: -plugin-opt=-: ld.lld: Unknown command line argument '-stack-alignment=8'.  Try 'ld.lld --help'
ld.lld: Did you mean '--stackrealign=8'?
```

**Comment by nickdesaulniers** [comment]
> let me find out what's going on, we should have had something somewhere fail

@samitolvanen pointed out the version string from CI is 2 days behind.

> clang version 13.0.0-++20210607111109+caf26612ddb5-1~exp1~20210607091856.4030

**Comment by torvic9** [comment]
> @torvic9 this would be a great kernel patch for you to send!
> 
> You might want to wait until https://reviews.llvm.org/D103928 is actually merged, but you can prep a patch here. The same patch should additionally modify the use of `-plugin-opt=-warn-stack-size=` in the top level Makefile conditional on lld-13 as well.

`-warn-stack-size` only appears in 5.13 but not in 5.12.
> 
> You can add link tags for:
> 
> ```
> Link: https://github.com/ClangBuiltLinux/linux/issues/1377
> Link: https://reviews.llvm.org/D52322
> Link: https://reviews.llvm.org/D103048
> Link: https://reviews.llvm.org/D103928
> ```
> 
> in the commit message, I would appreciate that.
> 
> I'm hoping to land https://reviews.llvm.org/D103928 today, so it would be good to have a kernel patch ready to go soon, since our CI is currently red without such kernel patch. (or it should be, let me find out what's going on, we should have had something somewhere fail).

Okay, I'll try to prepare a patch and submit it once the above has landed.
Should it be based on 5.12 or 5.13 (see above, warn-stack-size is not in 5.12)?
(Edit: I guess that I just do both)

**Comment by nickdesaulniers** [comment]
Kernel patch for `-plugin-opt=-code-model=kernel -plugin-opt=-stack-alignment=`:
https://lore.kernel.org/lkml/214134496.67043.1623317284090@office.mailbox.org/

**Comment by nickdesaulniers** [comment]
> Should it be based on 5.12 or 5.13

Always use mainline; patches land there first then eventually get back ported to stable (either via clean cherry picks, or manual backports if necessary).  Or rather than mainline, I prefer developing off of linux-next. Some maintainers are particular that you use branches of their trees (if what you're touching is under active development).

**Comment by torvic9** [comment]
Okay, I will submit it again. I hope I get it right this time... ;)

**Comment by nathanchance** [comment]
v2: https://lore.kernel.org/r/f2c018ee-5999-741e-58d4-e482d5246067@mailbox.org/

**Comment by nickdesaulniers** [comment]
https://reviews.llvm.org/D103928 has landed as https://reviews.llvm.org/rGfc018ebb608ee0c1239b405460e49f1835ab6175. @torvic9 would you mind sending a kernel patch for that too, please? :)

**Comment by torvic9** [comment]
Sure :)
Will this have a 'stable' tag as well? ~~Current stable kernel does not have '-warn-stack-size', it was introduced with this commit: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/Makefile?h=v5.13-rc5&id=24845dcb170e16b3100bd49743687648c71387ae~~

EDIT: I have just noticed that this is [now in the stable-queue](https://git.kernel.org/pub/scm/linux/kernel/git/stable/stable-queue.git/diff/queue-5.12/makefile-lto-have-linker-check-wframe-larger-than.patch?id=c99efadac42935e00ba9dcc08b9de5b8a567a15d) 

So I guess the patch also needs a "Fixes: <commit>" tag?

On a sidenote, shouldn't the LLVM commit message say
`make -warn-stack-size into a module attr`
instead of
`make -warn-frame-size into a module attr`?
___
(By the way, I highly appreciate that you and Nathan guided me through the patch submission process. A proper explanation makes things much easier to understand. Thanks!)

**Comment by torvic9** [comment]
Here is the diff with a small addition to the comment above:
```diff
diff --git a/Makefile b/Makefile
index 144d4f8b7eb3..51a1b6b2c2ab 100644
--- a/Makefile
+++ b/Makefile
@@ -929,11 +929,14 @@ CC_FLAGS_LTO	+= -fvisibility=hidden
 # Limit inlining across translation units to reduce binary size
 KBUILD_LDFLAGS += -mllvm -import-instr-limit=5
 
-# Check for frame size exceeding threshold during prolog/epilog insertion.
+# Check for frame size exceeding threshold during prolog/epilog insertion
+# when using lld < 13.0.0.
 ifneq ($(CONFIG_FRAME_WARN),0)
+ifeq ($(shell test $(CONFIG_LLD_VERSION) -lt 130000; echo $$?),0)
 KBUILD_LDFLAGS	+= -plugin-opt=-warn-stack-size=$(CONFIG_FRAME_WARN)
 endif
 endif
+endif
 
 ifdef CONFIG_LTO
 KBUILD_CFLAGS	+= -fno-lto $(CC_FLAGS_LTO)
```

**Comment by nickdesaulniers** [comment]
> Will this have a 'stable' tag as well?

Yes; we support LTO as of 5.12 which is stable: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/arch/Kconfig?h=v5.12.10#n656

> So I guess the patch also needs a "Fixes: " tag?

Yep. `Fixes: 24845dcb170e ("Makefile: LTO: have linker check -Wframe-larger-than")`

> shouldn't the LLVM commit message say

Yes; too late to change it now.

> I highly appreciate

De nada. Good job. Diff LGTM

**Comment by kraj** [comment]
@nickdesaulniers now that https://reviews.llvm.org/D103048 has landed in llvm, I think we need to open an issue with mesa as well to stop using it see [1]

[1] https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/gallium/auxiliary/gallivm/lp_bld_misc.cpp#L354

**Comment by nathanchance** [comment]
Both fixes merged into mainline (thanks @torvic9 and @kees!):

https://git.kernel.org/linus/2398ce80152aae33b9501ef54452e09e8e8d4262
https://git.kernel.org/linus/0236526d76b87c1dc2cbe3eb31ae29be5b0ca151

This should be resolved on both the kernel and the toolchain side so closing.
